### PR TITLE
disable and hide csv import button when request in flight

### LINKED ISSUE #63: Improved performance for large table refresh - Refresh single rows instead of entire table
After staging many users (eg. an entire county), the app exerpience slows down.  Uploads (bulk and single), refreshes, and page load (eg edit) take a long time.  Can we be more selective inwhat we are refreshing to make this experience faster?

### LINKED ISSUE #136: Duplicate requests when button is clicked quickly before state change
Reproducible when clicking `Add` to upload a csv and i believe this should also affect the `Upload` button

### PR DESCRIPTION
**Pull request by freddieptf**
Fixes #136 
Fixes #63

### CODE REVIEW

File: src/server.ts
Comment (kennsippell) [inline src/server.ts:25]: looks like it is recommended to enable this only in production environments. can we do that?
https://liquidjs.com/tutorials/use-in-expressjs.html#Caching

i personally like the dev experience of changing the view and having it reload without restarting the service

File: src/routes/add-place.ts
Comment (kennsippell) [inline src/routes/add-place.ts:78]: i don't think we should do this kinda thing. it is too easy to make a change in the progress-bar view and not know you need to update this. 

can we find another way? maybe use something like a view with `hx-swap-oob` instead?

File: src/liquid/place/bulk_create_form.html
Comment (kennsippell) [inline src/liquid/place/bulk_create_form.html:7]: i like the use of `hx-disable-alt` why do we need to do all the other stuff? are you trying to hide it?
is disabling the button not enough?

File: src/liquid/place/bulk_create_form.html
Comment (kennsippell) [inline src/liquid/place/bulk_create_form.html:115]: can you explain why you're doing this?

**Review by kennsippell — CHANGES_REQUESTED** [review]
good idea and good change

File: src/liquid/place/bulk_create_form.html
Comment (freddieptf) [reply src/liquid/place/bulk_create_form.html:7]: After uploading the csv there's a period of time, when fetching the next page, that the buttons are still visible. You can still click the `Add` button then. I think i can just use the progress bar here from the start and not use `hx-disabled-elt`

File: src/liquid/place/bulk_create_form.html
Comment (freddieptf) [reply src/liquid/place/bulk_create_form.html:115]: I am doing this so i can swap the `Add` view with the progress bar before initiating the redirect to the product list. For this i trigger a custom event that signals when we're ready to redirect. Before, the redirect happened immediately but since the place list takes some time to load, the bulk upload form will still be visible until the request for the the place list page completes

File: src/liquid/place/bulk_create_form.html
Comment (kennsippell) [reply src/liquid/place/bulk_create_form.html:7]: Is `hx-indicator` useful?
Could we maybe use an animated spinner (https://github.com/medic/cht-user-management/issues/62) instead making them empty?

File: src/liquid/place/bulk_create_form.html
Comment (kennsippell) [reply src/liquid/place/bulk_create_form.html:115]: Wonder if chunking the table response would be another option?  So rather than waiting for the whole table to be ready and the whole response to come in, we could send some minimal chunk.

In my mind, this might be a little better because it avoids the complexity of this custom event and the swapping complexity + it is a good practice for large pages

File: src/liquid/place/bulk_create_form.html
Comment (freddieptf) [reply src/liquid/place/bulk_create_form.html:7]: Maybe not, the indicator will be visible for the duration of the csv upload request. When that requests completes there's still the transition period between the upload form and the next page where a user could still click the `Add` button

File: src/liquid/place/bulk_create_form.html
Comment (freddieptf) [reply src/liquid/place/bulk_create_form.html:115]: Sounds good, we can definitely show the table then lazy load the data

Comment (freddieptf) [reply src/liquid/place/bulk_create_form.html]: I've implemented the change, first we load the skeleton table then load the items.

File: src/liquid/place/bulk_create_form.html
Comment (kennsippell) [reply src/liquid/place/bulk_create_form.html:7]: Is this still a problem now that we lazy-load the data?

Comment (freddieptf) [reply src/liquid/place/bulk_create_form.html:7]: I also think It's safe to ignore now, the user would need a really bad connection or sub-100ms reaction time to view/click that button

**Comment by kennsippell** [comment]
The performance tweaks in this PR are spot-on.

For generation times, measured the change to enable `cache: true` in LiquidJS and it brought down time-to-generate by 80% (2.5s down to 0.5s).  For response size, it brought responses sizes down 97% (from 3.5MB to ~125kb).

File: src/server.ts
Comment (kennsippell) [inline src/server.ts:28]: The performance tweaks in this PR are spot-on.

For generation times, measured the change to enable `cache: true` in LiquidJS and it brought down time-to-generate by 80% (2.5s down to 0.5s).  
For response size, compression brought responses sizes down 97% (from 3.5MB to ~125kb).

**Review by kennsippell — CHANGES_REQUESTED** [review]
Sorry! I didn't send the code review... I left in pending state

**Review by kennsippell — CHANGES_REQUESTED** [review]
I didn't see this review was pending. Apologies for the slowness.

* I just did a quick test and I think the buttons in the rows aren't working - try clicking "upload", "trash can", or "refresh" on any row in the table and nothing happens.
* During CSV upload could we maybe disable the cancel button too? It's clickable but I'm not sure if it does anything.

Code looks good!

**Comment by freddieptf** [comment]
> I just did a quick test and I think the buttons in the rows aren't working - try clicking "upload", "trash can", or "refresh" on any row in the table and nothing happens.

oh i missed that, i think it's a combination of how swaps are done on the table and https://github.com/bigskysoftware/htmx/pull/2247. I've updated the libraries and added support for individual row updates

**Comment by freddieptf** [comment]
@kennsippell I've made some changes to the events api and the updates should appear "faster" now. This should fix https://github.com/medic/cht-user-management/issues/63 as well

File: src/routes/app.ts
Comment (kennsippell) [inline src/routes/app.ts:91]: shouldn't this just be the directive.html view and let it route accordingly?

File: src/routes/events.ts
Comment (kennsippell) [inline src/routes/events.ts:40]:
```suggestion
            place,
```

**Review by kennsippell — APPROVED** [review]
Excellent change!
Remember to update package version before merge.

**Review by kennsippell — CHANGES_REQUESTED** [review]
I've got a problem when submitting invalid csv files. No error is displayed.

File: src/services/upload-manager.ts
Comment (freddieptf) [reply src/services/upload-manager.ts:98]: > when you edit CHP A you can cause changes to CHP B. I suppose I will now have to monitor all changes on all rows and separately trigger each refresh

Triggering individual events is the most optimal. Refreshing the table is basically just refreshing the page at this point

File: src/liquid/components/place_item.html
Comment (freddieptf) [reply src/liquid/components/place_item.html:66]: Anything after the new line will be ignored when pushing the html via sse https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format

**Review by kennsippell — APPROVED** [review]

**Pull request merged**
